src/models/StyledComponent.js

```diff
@@ -236,7 +238,27 @@ export default (ComponentStyle: Function) => {
         ? `${escape(options.displayName)}-${options.componentId}`
         : options.componentId || componentId
 
-    const componentStyle = new ComponentStyle(rules, attrs, styledComponentId)
+    let finalAttrs = attrs
+
+    // fold the underlying StyledComponent attrs up (implicit extend)
+    // $FlowFixMe
+    if (targetIsSC && target.attrs) finalAttrs = { ...target.attrs, ...attrs }
+
+    const componentStyle = new ComponentStyle(
+      targetIsSC
+        ? // fold the underlying StyledComponent rules up (implicit extend)
+          // $FlowFixMe
+          target.componentStyle.rules.concat(rules)
```

Comment:
Should this be moved to the `ComponentStyle` constructor itself? i.e. 

```js
new ComponentStyle(rules, attrs, id, child)
```

where the child is another `ComponentStyle`; This would also move the `attrs` merging, I guess? Not quite sure why we don't move the bulk of `buildExecutionContext` there anyway.

Reply:
Could go either way. Currently ComponentStyle isn't concerned with folding which I think is probably a good thing to separate concerns?

Reply:
Yea, I was thinking of a "combined" refactor of StyledComponent and ComponentStyle, but that can certainly wait :+1: